routes/runtime/github.rb

```diff
@@ -0,0 +1,128 @@
+# frozen_string_literal: true
+
+class CloverRuntime
+  hash_branch("github") do |r|
+    if (runner = GithubRunner[vm_id: @vm.id]).nil? || (repository = runner.repository).nil?
+      fail CloverError.new(400, "InvalidRequest", "invalid JWT format or claim in Authorization header")
+    end
+
+    # getCacheEntry
+    r.get "cache" do
+      fail CloverError.new(204, "NotFound", "No cache entry") if repository.access_key.nil?
+
+      keys, version = r.params["keys"]&.split(","), r.params["version"]
+      fail CloverError.new(400, "InvalidRequest", "Wrong parameters") if keys.nil? || keys.empty? || version.nil?
+
+      # Clients can send multiple keys; we return the first matching cache in
+      # incoming key order. The function `.min_by { keys.index(_1.key) }` helps
+      # us achieve this by ordering entries based on the index of key in the
+      # given order. If the same cache exists for both the head_branch and the
+      # default branch, we prioritize and return the cache for the head_branch.
+      # The part `(scopes.index(_1.scope) * keys.size)` assists in sorting the
+      # caches by scope, pushing entries for later scopes to the end of the
+      # list.
+      scopes = [runner.workflow_job&.dig("head_branch"), repository.default_branch].compact
+      entry = repository.cache_entries_dataset
+        .exclude(committed_at: nil)
+        .where(key: keys, version: version, scope: scopes).all
+        .min_by { keys.index(_1.key) + (scopes.index(_1.scope) * keys.size) }
+
+      fail CloverError.new(204, "NotFound", "No cache entry") if entry.nil?
+
+      entry.update(last_accessed_at: Time.now, last_accessed_by: runner.id)
+      signed_url = repository.url_presigner.presigned_url(:get_object, bucket: repository.bucket_name, key: entry.blob_key, expires_in: 900)
+
+      {
+        scope: entry.scope,
+        cacheKey: entry.key,
+        cacheVersion: entry.version,
+        creationTime: entry.created_at,
+        archiveLocation: signed_url
+      }
+    end
+
+    r.on "caches" do
+      # listCache
+      r.get true do
+        key = r.params["key"]
+        fail CloverError.new(204, "NotFound", "No cache entry") if key.nil?
+
+        scopes = [runner.workflow_job&.dig("head_branch"), repository.default_branch].compact
+        entries = repository.cache_entries_dataset
+          .exclude(committed_at: nil)
+          .where(key: key, scope: scopes).all
+
+        {
+          totalCount: entries.count,
+          artifactCaches: entries.map do
+            {
+              scope: _1.scope,
+              cacheKey: _1.key,
+              cacheVersion: _1.version,
+              creationTime: _1.created_at
+            }
+          end
+        }
+      end
+
+      repository.setup_blob_storage unless repository.access_key
+
+      # reserveCache
+      r.post true do
+        key = r.params["key"]
+        version = r.params["version"]
+        size = r.params["cacheSize"].to_i
```

Comment:
Should we limit the size and fail the request if it exceeds that limit?

Reply:
The client has a limit, but it's still good to double-check here. I added that. However, the client might bypass this somehow since it connects directly to the blob storage. ref: https://github.com/ubicloud/ubicloud/pull/1668#issuecomment-2163187241